packages/maskbook/src/plugins/ITO/hooks/usePoolState.ts

```diff
@@ -0,0 +1,73 @@
+import { uniqBy } from 'lodash-es'
+import { useReducer } from 'react'
+import { isSameAddress } from '../../../web3/helpers'
+import type { EtherTokenDetailed, ERC20TokenDetailed } from '../../../web3/types'
+
+export interface PoolState {
+    sellToken: EtherTokenDetailed | ERC20TokenDetailed
+    buyTokens: (EtherTokenDetailed | ERC20TokenDetailed)[]
+}
+
+export enum PoolActionType {
+    UPDATE_SELL_TOKEN,
+    ADD_BUY_TOKEN,
+    UPDATE_BUY_TOKEN,
+    REMOVE_BUY_TOKEN,
+}
+
+export type PoolAction =
+    | {
+          type: PoolActionType.UPDATE_SELL_TOKEN
+          token: PartialRequired<EtherTokenDetailed | ERC20TokenDetailed, 'address'>
+      }
+    | {
+          type: PoolActionType.ADD_BUY_TOKEN
+          token: EtherTokenDetailed | ERC20TokenDetailed
+      }
+    | {
+          type: PoolActionType.REMOVE_BUY_TOKEN
+          token: PartialRequired<EtherTokenDetailed | ERC20TokenDetailed, 'address'>
+      }
+    | {
+          type: PoolActionType.UPDATE_BUY_TOKEN
+          token: PartialRequired<EtherTokenDetailed | ERC20TokenDetailed, 'address'>
+      }
+
+function reducer(state: PoolState, action: PoolAction): PoolState {
+    switch (action.type) {
+        case PoolActionType.UPDATE_SELL_TOKEN:
+            return {
+                ...state,
+                sellToken: {
```

Comment:
you can use `immer` (in our deps) to simplify this code

Reply:
ok